### ISSUE TITLE
Reflection for records fails with C# 9.0

### ISSUE THREAD

**Issue by rmihael** [fscheck/FsCheck#564]
In C# 9.0 compiler generates properties of `record` classes as init-only. It appears that `isCSharpRecordType` does not accommodate for this and fails to recognize such classes as records. To be more specific, reflection-based generation fails for this:

```
public record RgbColor(byte Red, byte Green, byte Blue) {}
```

This works:

```
public record RgbColor(byte Red, byte Green, byte Blue)
{
    public byte Red { get; } = Red;

    public byte Green { get; } = Green;

    public byte Blue { get; } = Blue;
}
```

This post suggests the way to detect init-only properties through reflection: https://alistairevans.co.uk/2020/11/01/detecting-init-only-properties-with-reflection-in-c-9/. I believe that `isCSharpRecordType` must be fixed.

**Comment by kurtschelfthout** [comment]
Sounds good, thanks for investigating. Can you send a PR for 💯 karma points? ;)

**Comment by rmihael** [comment]
I will work on it during this week.

**Comment by kurtschelfthout** [comment]
Awesome, thanks in advance!

**Comment by rmihael** [comment]
@kurtschelfthout while working on integrating FsCheck into my project, I stumbled upon another problem: https://stackoverflow.com/questions/67195933/scoping-contant-fscheck-generator-to-each-test. This GitHub issue is not the best place to ask for help for something unrelated, but may I humbly ask if you can take a look? :)

**Comment by kurtschelfthout** [comment]
@rmihael I left a comment - I do monitor fscheck tagged questions on stackoverflow so would've answered eventually (though notifications are often delayed by a day or so).

**Comment by rmihael** [comment]
I tried my best to prepare the pull request, but it turned out much more involved than I expected. The `IsExternalInit` marker class is only available in the NET5.0 framework. There is a well-known workaround for this for older versions (https://btburnett.com/csharp/2020/12/11/csharp-9-records-and-init-only-setters-without-dotnet5.html), but I have not managed to make it work with F#. 
I do not know if you plan to restrict newer versions of FsCheck to NET5.0+ framework versions. I tried to do it out of curiosity, but it was not trivial for someone unfamiliar with the F# ecosystem. At the very least, Packet has to be updated to 6.0 which is still in beta and gifted me with a bit of old-times nostalgia upon run:

```
         =================================================================
         	Basic Fault Address Reporting
         =================================================================
         Memory around native instruction pointer (0x109e55baf):0x109e55b9f  17 00 49 89 c7 31 db 4d 85 ff 0f 84 f5 00 00 00  ..I..1.M........
         0x109e55baf  45 0f b7 07 41 81 f8 4d 5a 00 00 75 33 44 8b 45  E...A..MZ..u3D.E
         0x109e55bbf  d4 41 8b 77 3c 48 8d 86 f8 00 00 00 49 39 c0 73  .A.w<H......I9.s
         0x109e55bcf  2b 48 8d 15 fc c1 1b 00 48 8d 0d dc c1 1b 00 31  +H......H......1```

**Comment by kurtschelfthout** [comment]
Why does `IsExternalInit` need to be referenced by FsCheck? Perhaps I'm missing something, but isn't it a matter of checking for `IsExternalInit` (which you can do by name via reflection, so no real reference needed) and then making sure that the reflection-based generator can create that type via reflection somehow?

If this is for testing, yes that is perhaps a bit tricky since FsCheck.Test is F#. Best is probably to reference a .NET 5 C# project with test types; but if that is too involved I'd be happy with an example in one of the C# projects in the examples folder (which you can update to .NET 5)

**Comment by AlexeyRaga** [comment]
Stumbled upon the same issue. Are there still ideas on how it can be handled? 
I can probably find some time to dedicate if there is a direction to go...

**Comment by kurtschelfthout** [comment]
Lots of build gymnastics necessary for this unfortunately. I think I got something together in the attached PR though, it's a failing test at least :)

**Comment by AlexeyRaga** [comment]
@kurtschelfthout thanks for working in this!

**Comment by kurtschelfthout** [comment]
Released in 2.16.0.

**Comment by dgrozenok** [comment]
Could it please be merged to the 3.0.0-alpha?

**Comment by kurtschelfthout** [comment]
@dgrozenok yes will do at some point. There's loads of other API changes in the unreleased 3.0 branch though, so it won't be a smooth upgrade.

**Comment by kurtschelfthout** [comment]
Released in 3.0.0-beta1, along with a bunch of other backwards incompatible changes. Please read the release notes before upgrading from 3.0.0-alpha.
